include/oneapi/dpl/internal/dynamic_selection_impl/auto_tune_policy.h

```diff
@@ -0,0 +1,251 @@
+        } else if (resample_time_ == never_resample) {
+          return use_best_resource;
+        } else {
+          auto now = std::chrono::high_resolution_clock::now();
+          auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(now-t0_).count();
+          if (ms < resample_time_) {
+            return use_best_resource;
+          } else {
+            t0_ = now;
+            return next_resource_to_profile_ = 0;
+          }
+        }
+      }
+
+      // called to add new profile info
+      void add_new_timing(resource_with_offset_t r, timing_t t) {
+        std::unique_lock<std::mutex> l(m_);
+        auto offset = r.offset_;
+        timing_t new_value = t;
+        if (time_by_offset_.count(offset) == 0) {
+          time_by_offset_[offset] = time_data_t{1, t};
+        } else {
+          auto &td = time_by_offset_[offset];
+          auto n = td.num_timings_;
+          new_value = (n*td.value_+t)/(n+1);
+          td.num_timings_ = n+1;
+          td.value_ = new_value;
+        }
+        if (new_value < best_timing_) {
+          best_timing_ = new_value;
+          best_resource_ = r;
+        }
+      } 
+
+    };
+
+   public:
+
+    // Needed by Policy Traits
+    using resource_type = decltype(unwrap(std::declval<wrapped_resource_t>()));
+    using wait_type = typename Backend::wait_type;
+    using selection_type = internal::auto_tune_selection_type<Backend, resource_with_offset_t, tuner_t, KeyArgs...>;
+
+    auto_tune_policy(double resample_time=never_resample) {
+      if (resample_time != deferred_initialization) {
+        initialize(resample_time);
+      }
+    }
+
+    auto_tune_policy(const std::vector<resource_type>& u, double resample_time=never_resample) {
+      if (resample_time != deferred_initialization) {
+        initialize(u, resample_time);
+      }
+    }
+   
+    void initialize(double resample_time=never_resample) {
+      backend_ = std::make_shared<Backend>();
+      initialize_impl(resample_time);
+    }
+
+    void initialize(const std::vector<resource_type>& u, double resample_time=never_resample) {
+      backend_ = std::make_shared<Backend>(u);
+      initialize_impl(resample_time);
+    }
+
+    template<typename Function, typename ...Args>
+    selection_type select(Function&& f, Args&&...args) {
+      auto k =  make_task_key(std::forward<Function>(f), std::forward<Args>(args)...);
+      auto t  = tuner_by_key_[k];
+      auto offset = t->get_resource_to_profile();
+      if (offset == use_best_resource) {
+        return selection_type{*this, t->best_resource_, t}; 
+      } else {
+        auto r = resources_with_offset_[offset];
+        return selection_type{*this, r, t}; 
+      } 
+    }
+
+    template<typename Function, typename ...Args>
+    auto submit(selection_type e, Function&& f, Args&&... args) {
+      return backend_->submit(e, std::forward<Function>(f), std::forward<Args>(args)...);
+    }
+
+    auto get_resources() {
+       if (backend_) {
```

Comment:
Should we be checking for the size of the resources instead?

Reply:
At least in auto_tune, this is a test for deferred_initialization. No backend_ is created if deferred and not yet initialized.